File: pandiv2env.lua
function Div(elem)
    if FORMAT ~= 'latex' then return end
    local attr = elem.attributes["data-environment"]
    if attr then
        -- Find the first ','
        local _, loc = string.find(attr, ',')
        -- If we have no arguments, the full string is the environment
        if not loc then loc = string.len(attr) + 1 end
        --print(attr, loc)

        -- Now split into the environment and arguments
        local env = string.sub(attr, 1, loc-1)
        local args = string.sub(attr, loc+1, string.len(attr)+1)

        -- If we have arguments, put them inside '[...]'
        if args ~= ""  then args = "[" .. args .. "]" end
        --print("env = '" .. env .. "'", "args = '" .. args .. "'")

        -- Now create the barriers.  We include the '%' to strip the
        -- intervening spaces
        local begin = pandoc.RawBlock(
                FORMAT, "\\begin{" .. env .. "}" .. args .. "%"
            )
        --for k,v in pairs(elem.content) do print(k,v) end
        local last = pandoc.RawInline(FORMAT, "\\end{" .. env .. "}")
        elem.content:extend({last})
        --local div = pandoc.Div(elem.content:extend({last}))
        return elem -- { begin, elem, last }
    end
end

File: pandiv2env.py
#!/usr/bin/env python

import panflute

def div2env(elem, doc, debug=False):
    if type(elem) == panflute.Div or doc.format in ["latex", "beamer"]:
        attr = getattr(elem, "attributes", None)
        if not attr:
            return

        env = attr.get("data-environment", "").split(",")
        if not env[0]:
            # This is not marked as an environment
            return

        # Convert the positional arguments to the proper \LaTeX format
        args = attr.get("data-environment-args", "").split(",")
        args = "{{{0}}}".format("}{".join(args)) if args[0] else ""

        # Enclose the keyword arguments in '[...]'
        opt = attr.get("data-environment-keyword", "")
        opt = "[{0}]".format(opt) if opt else ""

        begin = panflute.RawInline(
                "\\begin{{{0}}}{1}{2}%\n".format(env[0], opt, args),
                format="latex"
            )
        end = panflute.RawInline(
                "%\n\\end{{{0}}}".format(env[0]),
                format="latex"
            )
        if debug:
            panflute.debug("begin = '{0!s}'".format(begin))
            panflute.debug("end = '{0!s}'".format(end))

        elem.content[0] = panflute.Para(begin, *elem.content[0].content)
        elem.content[-1] = panflute.Para(*elem.content[-1].content, end)
        if debug:
            panflute.debug("content = '{0!s}'".format(elem.content[0]))

        return elem

    return

def main(doc=None):
    return panflute.run_filter(div2env, doc=doc)

if __name__ == "__main__":
    main()

